This notebook performs data cleaning for the scraped DataFrame.

## Remove missing rows

In [6176]:
import pandas as pd

In [6177]:
df = pd.read_csv("scraped_police_reports14.csv") # dataframe of all scraped articles
df.head()

name  \
0            Andrew Allen, badge #37   
1          Elijah Allen, badge #None   
2      Michael Anderson, badge #None   
3  Oluwademilade Adediran, badge #21   
4         Wade Anderson, badge #None   

                                 department  \
0  Department:Minneapolis Police Department   
1         Department:Minnesota State Patrol   
2  Department:Martin County Sheriffs Office   
3  Department:Minneapolis Police Department   
4       Department:Winona Police Department   

                                                 url  \
0  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1                                  No articles found   
2                                  No articles found   
3                                  No articles found   
4                                  No articles found   

                                                text  
0  Home Legislative File 2021-01132 RCA Legal s...  
1                                  No articles found  
2                                  No articles found  
3                                  No articles found  
4                                  No articles found

In [6178]:
df.shape # There are 15,527 articles (or links without articles)

(15527, 4)

In [6179]:
df.dtypes

name          object
department    object
url           object
text          object
dtype: object

In [6180]:
df.isna().sum() # There are 205 missing articles

name            0
department      0
url             0
text          205
dtype: int64

In [6181]:
null_text_rows = df.loc[df["text"].isna()] # Find the rows where there is no article text
null_text_rows.head()

name  \
31        Roxanne Affeldt, badge #None   
645          Todd Babekuhl, badge #246   
657          Todd Babekuhl, badge #246   
719  Stephanie L. Bailey, badge #29300   
726         Daniel Anderson, badge #94   

                                   department  \
31      Department:Champlin Police Department   
645  Department:Minneapolis Police Department   
657  Department:Minneapolis Police Department   
719     Department:St. Paul Police Department   
726  Department:Minneapolis Police Department   

                                                   url text  
31   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  NaN  
645  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  NaN  
657  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  NaN  
719  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  NaN  
726  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  NaN

We will drop the rows where `text` is `NaN`, because that means there is no article for that row.

In [6182]:
df = df.dropna() # Drop rows where there is any NaN value
df.shape

(15322, 4)

In [6183]:
len(df["url"].unique()) # There are 1855 unique URLs in the dataset and over 15000 total rows. So, many articles come from the same URL

1855

Below, we find the rows where the text or the URL, respectively, is equal to "No articles found".

In [6184]:
no_article_rows = df.loc[df["url"] == "No articles found"]
no_article_rows

name  \
1                  Elijah Allen, badge #None   
2              Michael Anderson, badge #None   
3          Oluwademilade Adediran, badge #21   
4                 Wade Anderson, badge #None   
6      Fernando F. Abla-Reyes, badge #801160   
...                                      ...   
15483             John Zupancic, badge #None   
15484               Brian Zwach, badge #None   
15485              Kenneth Zwak, badge #None   
15486        Christian Zweerink, badge #None   
15487            Kevin Zydowsky, badge #None   

                                          department                url  \
1                  Department:Minnesota State Patrol  No articles found   
2           Department:Martin County Sheriffs Office  No articles found   
3           Department:Minneapolis Police Department  No articles found   
4                Department:Winona Police Department  No articles found   
6              Department:St. Paul Police Department  No articles found   
...                                              ...                ...   
15483           Department:Gilbert Police Department  No articles found   
15484  Department:Oak Park Heights Police Department  No articles found   
15485            Department:Duluth Police Department  No articles found   
15486       Department:Park Rapids Police Department  No articles found   
15487       Department:Carver County Sheriffs Office  No articles found   

                    text  
1      No articles found  
2      No articles found  
3      No articles found  
4      No articles found  
6      No articles found  
...                  ...  
15483  No articles found  
15484  No articles found  
15485  No articles found  
15486  No articles found  
15487  No articles found  

[12418 rows x 4 columns]

In [6185]:
no_text_rows = df.loc[df["text"] == "No articles found"]
no_text_rows

name  \
1                  Elijah Allen, badge #None   
2              Michael Anderson, badge #None   
3          Oluwademilade Adediran, badge #21   
4                 Wade Anderson, badge #None   
6      Fernando F. Abla-Reyes, badge #801160   
...                                      ...   
15483             John Zupancic, badge #None   
15484               Brian Zwach, badge #None   
15485              Kenneth Zwak, badge #None   
15486        Christian Zweerink, badge #None   
15487            Kevin Zydowsky, badge #None   

                                          department                url  \
1                  Department:Minnesota State Patrol  No articles found   
2           Department:Martin County Sheriffs Office  No articles found   
3           Department:Minneapolis Police Department  No articles found   
4                Department:Winona Police Department  No articles found   
6              Department:St. Paul Police Department  No articles found   
...                                              ...                ...   
15483           Department:Gilbert Police Department  No articles found   
15484  Department:Oak Park Heights Police Department  No articles found   
15485            Department:Duluth Police Department  No articles found   
15486       Department:Park Rapids Police Department  No articles found   
15487       Department:Carver County Sheriffs Office  No articles found   

                    text  
1      No articles found  
2      No articles found  
3      No articles found  
4      No articles found  
6      No articles found  
...                  ...  
15483  No articles found  
15484  No articles found  
15485  No articles found  
15486  No articles found  
15487  No articles found  

[12418 rows x 4 columns]

In [6186]:
no_text_rows.equals(no_article_rows) # Check if the rows without text and the rows without articles are the same

True

There are 12,418 rows with no article text, out of about 15,000 total rows. We will drop the rows with no article text.

In [6187]:
df = df.loc[(df["url"] != "No articles found") & (df["text"] != "No articles found")]
df

name  \
0          Andrew Allen, badge #37   
5      Guled Abdullahi, badge #706   
11     Dean V. Albers, badge #None   
22         Scott Aikins, badge #22   
24       Matthew Aish, badge #None   
...                            ...   
15520      David Voss, badge #7441   
15522  Roderick Weber, badge #7612   
15524  Roderick Weber, badge #7612   
15525  Roderick Weber, badge #7612   
15526   Kent Warnberg, badge #7570   

                                            department  \
0             Department:Minneapolis Police Department   
5      Department:Hennepin County Sheriff's Department   
11      Department:Goodhue County Sheriff's Department   
22            Department:Minneapolis Police Department   
24       Department:Columbia Heights Police Department   
...                                                ...   
15520         Department:Minneapolis Police Department   
15522         Department:Minneapolis Police Department   
15524         Department:Minneapolis Police Department   
15525         Department:Minneapolis Police Department   
15526         Department:Minneapolis Police Department   

                                                     url  \
0      https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
5      https://assets.nationbuilder.com/cuapb/pages/1...   
11     https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
22     https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
24     https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
...                                                  ...   
15520  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
15522  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
15524  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
15525  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
15526  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   

                                                    text  
0      Home Legislative File 2021-01132 RCA Legal s...  
5      Hennepin County  300 South Sixth Street, Minne...  
11     2/22/2021 Jenson v. Craft, Civil No. 01-1488(D...  
22     2/22/2021 United States v. Diriye, Case No. 14...  
24     Too_Long1    Arbitration   LELS (Mathew Aish)/...  
...                                                  ...  
15520  MINNEAPOLIS POLICE DEPARTMENT  Internal Affair...  
15522                                           Too_Long  
15524                                      No text found  
15525                                           Too_Long  
15526  they were both on break at the Holiday gas sta...  

[2904 rows x 4 columns]

Now investigate rows where `text` is equal to `Too_Long` or `No text found`

In [6188]:
df.loc[df["text"] == "Too_Long"]

name  \
928         Tyrone Barze, badge #330   
929         Tyrone Barze, badge #330   
1060        Tyrone Barze, badge #330   
1135        Tyrone Barze, badge #330   
1346   Mukhtar Abdulkadir, badge #47   
...                              ...   
15271    Roderick Weber, badge #7612   
15276    Roderick Weber, badge #7612   
15516    Roderick Weber, badge #7612   
15522    Roderick Weber, badge #7612   
15525    Roderick Weber, badge #7612   

                                     department  \
928    Department:Minneapolis Police Department   
929    Department:Minneapolis Police Department   
1060   Department:Minneapolis Police Department   
1135   Department:Minneapolis Police Department   
1346   Department:Minneapolis Police Department   
...                                         ...   
15271  Department:Minneapolis Police Department   
15276  Department:Minneapolis Police Department   
15516  Department:Minneapolis Police Department   
15522  Department:Minneapolis Police Department   
15525  Department:Minneapolis Police Department   

                                                     url      text  
928    https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Too_Long  
929    https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Too_Long  
1060   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Too_Long  
1135   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Too_Long  
1346   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Too_Long  
...                                                  ...       ...  
15271  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Too_Long  
15276  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Too_Long  
15516  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Too_Long  
15522  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Too_Long  
15525  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Too_Long  

[96 rows x 4 columns]

In [6189]:
df.loc[df["text"] == "No text found"]

name  \
25      Charles Adams, III, badge #59   
98        Dean V. Albers, badge #None   
130     Mukhtar Abdulkadir, badge #47   
161          Lance Ahlrich, badge #18   
192        Daniel Anderson, badge #94   
...                               ...   
15490  Rhonda Ziegelmann, badge #None   
15496  Rhonda Ziegelmann, badge #None   
15507     Roderick Weber, badge #7612   
15513   Geoffrey Toscano, badge #7257   
15524     Roderick Weber, badge #7612   

                                            department  \
25            Department:Minneapolis Police Department   
98      Department:Goodhue County Sheriff's Department   
130           Department:Minneapolis Police Department   
161           Department:Minneapolis Police Department   
192           Department:Minneapolis Police Department   
...                                                ...   
15490  Department:Hennepin County Sheriff's Department   
15496  Department:Hennepin County Sheriff's Department   
15507         Department:Minneapolis Police Department   
15513         Department:Minneapolis Police Department   
15524         Department:Minneapolis Police Department   

                                                     url           text  
25     https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  No text found  
98     https://www.dropbox.com/s/skcxonxtqyo7vi2/POST...  No text found  
130    https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  No text found  
161    https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  No text found  
192    https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  No text found  
...                                                  ...            ...  
15490  https://assets.nationbuilder.com/cuapb/pages/1...  No text found  
15496  https://assets.nationbuilder.com/cuapb/pages/1...  No text found  
15507  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  No text found  
15513  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  No text found  
15524  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  No text found  

[400 rows x 4 columns]

There are 400 rows where the text is too long and 96 rows with no text found. We will drop these rows. That results in a DataFrame with 2,408 rows.

In [6190]:
df = df.loc[~df["text"].isin(["No text found", "Too_Long"])] # Rows where the text is not equal to "No text found" AND not equal to "Too_Long".
df = df.reset_index(drop=True) # reset index to a sequential integer index
df

name  \
0               Andrew Allen, badge #37   
1           Guled Abdullahi, badge #706   
2           Dean V. Albers, badge #None   
3               Scott Aikins, badge #22   
4             Matthew Aish, badge #None   
...                                 ...   
2403  Richard D. Zimmerman, badge #7961   
2404      Michael N. Young, badge #7895   
2405        Steve Wuorinen, badge #7876   
2406            David Voss, badge #7441   
2407         Kent Warnberg, badge #7570   

                                           department  \
0            Department:Minneapolis Police Department   
1     Department:Hennepin County Sheriff's Department   
2      Department:Goodhue County Sheriff's Department   
3            Department:Minneapolis Police Department   
4       Department:Columbia Heights Police Department   
...                                               ...   
2403         Department:Minneapolis Police Department   
2404         Department:Minneapolis Police Department   
2405         Department:Minneapolis Police Department   
2406         Department:Minneapolis Police Department   
2407         Department:Minneapolis Police Department   

                                                    url  \
0     https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1     https://assets.nationbuilder.com/cuapb/pages/1...   
2     https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
3     https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
4     https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
...                                                 ...   
2403  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
2404  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
2405  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
2406  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
2407  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   

                                                   text  
0     Home Legislative File 2021-01132 RCA Legal s...  
1     Hennepin County  300 South Sixth Street, Minne...  
2     2/22/2021 Jenson v. Craft, Civil No. 01-1488(D...  
3     2/22/2021 United States v. Diriye, Case No. 14...  
4     Too_Long1    Arbitration   LELS (Mathew Aish)/...  
...                                                 ...  
2403  MINNEAPOLIS POLICE DEPARTMENT  INTERNAL AFFAIR...  
2404  Cook, Joni  ,:- --Irom:  Glampe, Travis  ..ien...  
2405  4/11/2021 Cop Gets $585K After Colleagues Snoo...  
2406  MINNEAPOLIS POLICE DEPARTMENT  Internal Affair...  
2407  they were both on break at the Holiday gas sta...  

[2408 rows x 4 columns]

## Check for valid values

Some articles in the `text` column start with `"Too_Long"`, even if there is article text. Let's examine those.

In [6191]:
too_long_articles = df.loc[df["text"].str.startswith("Too_Long")].head(20)
too_long_articles

name  \
4            Matthew Aish, badge #None   
10     Vincent A. Adams, badge #720505   
12       Philip Alejandrino, badge #29   
16       Philip Alejandrino, badge #29   
23         Christopher Abbas, badge #2   
37       Mukhtar Abdulkadir, badge #47   
43       Richard C. Altonen, badge #66   
78            Abdiwahab Ali, badge #30   
105      Medaria Arradondo, badge #231   
110  Carlos Baires Escobar, badge #275   
124       Richard Altonen, badge #2287   
127  Carlos Baires Escobar, badge #275   
131          Ryan Atkinson, badge #244   
138    Brandon Bartholomew, badge #324   
155           Tyrone Barze, badge #330   
157    Shannon L. Barnette, badge #268   
168         James F Archer, badge #177   
205        Katie Blackwell, badge #664   
208      Ashley Bergersen, badge #3302   
210            Heath Bird, badge #None   

                                        department  \
4    Department:Columbia Heights Police Department   
10           Department:St. Paul Police Department   
12        Department:Minneapolis Police Department   
16        Department:Minneapolis Police Department   
23        Department:Minneapolis Police Department   
37        Department:Minneapolis Police Department   
43        Department:Minneapolis Police Department   
78        Department:Minneapolis Police Department   
105       Department:Minneapolis Police Department   
110       Department:Minneapolis Police Department   
124       Department:Minneapolis Police Department   
127       Department:Minneapolis Police Department   
131       Department:Minneapolis Police Department   
138       Department:Minneapolis Police Department   
155       Department:Minneapolis Police Department   
157       Department:Minneapolis Police Department   
168       Department:Minneapolis Police Department   
205       Department:Minneapolis Police Department   
208       Department:Minneapolis Police Department   
210          Department:Plymouth Police Department   

                                                   url  \
4    https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
10   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
12   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
16   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
23   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
37   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
43   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
78   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
105  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
110  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
124  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
127  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
131  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
138  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
155  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
157  https://assets.nationbuilder.com/cuapb/pages/1...   
168  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
205  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
208  https://assets.nationbuilder.com/cuapb/pages/1...   
210  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   

                                                  text  
4    Too_Long1    Arbitration   LELS (Mathew Aish)/...  
10   Too_LongOFFICE OF THE RAMSEY COUNTY ATTORNEY J...  
12   Too_Long    UNITED STATES DISTRICT COURT  DIST...  
16   Too_LongUNITED STATES DISTRICT COURT  DISTRICT...  
23   Too_Long      UNITED STATES DISTRICT COURT    ...  
37   Too_Long  1 IN THE MATTER OF ARBITRATION BETWE...  
43   Too_Long2/25/2021 The Bad Cops: How Minneapoli...  
78   Too_LongOPCR Case #17-03832  Table of Contents...  
105  Too_Long2/25/2021 The Bad Cops: How Minneapoli...  
110  Too_LongUNITED STATES DISTRICT COURT  DISTRICT...  
124  Too_Long2/25/2021 The Bad Cops: How Minneapoli...  
127  Too_Long    UNITED STATES DISTRICT COURT  DIST...  
131  Too_Long1364417 1364418 1364419 1364420 136442...  

Below, we examine the articles that start with `"Too_Long"`.

In [6192]:
too_long_articles.loc[155, "text"] # investigate one of the too_long rows

'Too_Long2/25/2021 The Bad Cops: How Minneapolis protects its worst police officers until it’s too late - Minnesota Reformer https://minnesotareformer.com/2020/12/15/the-bad-cops-how-minneapolis-protects-its-worst-police-officers-until-its-too-late/ 1/24 The Bad Cops: How Minneapolis protects its worst police o¨cers until it’s too late OFFICER BLAYNE LEHNER opened the back door of his squad car and grabbed the handcuffed teenager kicking the partition. He ripped Luis Garcia’s 5-foot-5, 125-pound body from the car and dropped him face down on the concrete. Garcia remembers feeling blows to his face before he lost consciousness. Three years later, when the Minneapolis Police Department was investigating the events of that night in December 2013, his sister Miriam Garcia said she heard loud noises like “someone hitting something metal.” She got out of the car she was waiting in to find her little brother motionless on the pavement, blood pooling beneath his face as five or six cops stood 

In [6193]:
too_long_articles.loc[157, "text"]

'Too_LongMinneapolis  Office of Police  Conduct Review  INVESTIGATIVE REPORT  Complaint Number: 17-10527  Investigator:  Stephen J McKean  Officer (s):  Shannon Barnette  Case Type:  Administrative  Date of Incident:  May 25, 2017  Complaint Filed:  June 07, 2017  Minneapolis  Police  13.43 - Personnel Data  CASE OVERVIEW  On 05/25/2017, at 18:11 hours, a 9-1-1 call was received from an unknown male subject who stated that  he was the Complainant\'s 13.82 however, he wished to remain anonymous. The caller asked for officers  to respond to Privacy Policy home for "a safety or a welfare check" as she has "mental health issues", the  nature of which he could not elaborate. 1 The caller stated that Privacy Policy called him "threatening my  family and me." The caller said that he didn\'t think she was going to do anything.2 He also stated, "I  mean, I don\'t think she\'s gonna hurt anyone, but..."3 At 20:14 hours on the 25th, Officer  13.43  13.43 and Officer  13.43  13.43 responded to  PP

In [6194]:
df.loc[2405, "text"] # examine a regular article to compare it with the "Too_Long" ones

'4/11/2021 Cop Gets $585K After Colleagues Snooped on Her DMV Data | WIRED https://www.wired.com/story/minnesota-police-dmv-database-abuse/ 1/5 LOUISE MATSAKIS SECURITY 06.21.2019 03:47 PM Minnesota Cop Awarded $585K After Colleagues Snooped on Her DMV Data A jury this week finds that Minneapolis police officers abused their license database access. Dozens of other lawsuits have made similar claims. IN 2013, AMY Krekelberg received an unsettling notice from Minnesota’s Department of Natural Resources: An employee had abused his access to a government driver’s license database and snooped on thousands of people in the state, mostly women. Krekelberg learned that she was one of them. MARLIN LEVISON/STAR TRIBUNE/GETTY IMAGES 4/11/2021 Cop Gets $585K After Colleagues Snooped on Her DMV Data | WIRED https://www.wired.com/story/minnesota-police-dmv-database-abuse/ 2/5 When Krekelberg asked for an audit of accesses to her DMV records, as allowed by Minnesota state law, she learned that her in

In [6195]:
len(too_long_articles), len(df)

(20, 2408)

There is still article text in the `"Too_Long"` articles, but it seems to be cut off in the middle of the sentence, which suggests that only a certain number of characters could be read before the article becomes too long. Just in case, we will keep the `"Too_Long"` at the beginning of these article texts, so that we know which articles might be incomplete. Besides, there are only 20 rows where the text is too long, out of 2,408 total rows.

Now we will check for any invalid values or empty strings in all columns

### Initial check of all columns

In [6196]:
# Check if any column contains an entry with all whitespace characters. It does not.
for col in df.columns:
    print(df[col].str.isspace().sum())

0
0
0
0


In [6197]:
# Check if any column contains an entry with all numeric characters. It does not
for col in df.columns:
    print(df[col].str.isnumeric().sum())

0
0
0
0


In [6198]:
# Remove leading and trailing whitespace from all entries
for col in df.columns: 
    df[col] = df[col].str.strip() # remove whitespace from that column

### Check invalid URLs

In [6199]:
df.sort_values(by = "url")

name  \
2258        Kent Warnberg, badge #7570   
1702       Jeffrey Pennaz, badge #5551   
1777           Kurt Radke, badge #5882   
2120      Craig A. Taylor, badge #7139   
2121          Cory Taylor, badge #7141   
...                                ...   
816         Mark Hanneman, badge #2654   
1113       Joseph Klimmek, badge #3735   
68         Tobias Anderson, badge #132   
1426        Daryl McKenna, badge #4623   
1413  Timothy J. Mattsson, badge #4490   

                                    department  \
2258  Department:Minneapolis Police Department   
1702  Department:Minneapolis Police Department   
1777  Department:Minneapolis Police Department   
2120  Department:Minneapolis Police Department   
2121  Department:Minneapolis Police Department   
...                                        ...   
816   Department:Minneapolis Police Department   
1113  Department:Minneapolis Police Department   
68    Department:Minneapolis Police Department   
1426  Department:Minneapolis Police Department   
1413  Department:Minneapolis Police Department   

                                                    url  \
2258  file:///home/chronos/u-6ec0973dcaad89553a5ddf1...   
1702  https://assets.nationbuilder.com/cuapb/pages/1...   
1777  https://assets.nationbuilder.com/cuapb/pages/1...   
2120  https://assets.nationbuilder.com/cuapb/pages/1...   
2121  https://assets.nationbuilder.com/cuapb/pages/1...   
...                                                 ...   
816   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1113  https://www.govinfo.gov/content/pkg/USCOURTS-m...   
68    https://www.govinfo.gov/content/pkg/USCOURTS-m...   
1426  https://www.govinfo.gov/content/pkg/USCOURTS-m...   
1413  https://www.govinfo.gov/content/pkg/USCOURTS-m...   

                                                   text  
2258  Failed: No connection adapters were found for ...  
1702  Home Legislative File 2022-00234 RCA Legal S...  
1777  Home Legislative File 2022-00241 RCA Legal S...  
2120  Home Legislative File 2022-00233 RCA Legal S...  
2121  Home Legislative File 2022-00240 RCA Legal S...  
...                                                 ...  
816   MPD officer who killed Amir Locke disciplined ...  
1113  UNITED STATES DISTRICT COURT  DISTRICT OF MINN...  
68    UNITED STATES DISTRICT COURT  DISTRICT OF MINN...  
1426  UNITED STATES DISTRICT COURT DISTRICT OF MINNE...  
1413  UNITED STATES DISTRICT COURT DISTRICT OF MINNE...  

[2408 rows x 4 columns]

In [6200]:
print(f"URL: {df.loc[2258, "url"]}")
print(f"Text: {df.loc[2258, "text"]}")

URL: file:///home/chronos/u-6ec0973dcaad89553a5ddf1372e2bf8e9d7aa203/MyFiles/Downloads/Star_Tribune_Wed__May_5__1993_.pdf
Text: Failed: No connection adapters were found for 'file:///home/chronos/u-6ec0973dcaad89553a5ddf1372e2bf8e9d7aa203/MyFiles/Downloads/Star_Tribune_Wed__May_5__1993_.pdf'


We will drop row 2258 since it has an invalid URL (URLs should start with HTTPS), and the text is just an error message.

In [6201]:
df.drop(index=2258, inplace=True)

In [6202]:
unique_urls = df["url"].value_counts() # get unique values of URL
unique_urls

url
https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pages/270/attachments/original/1618168204/Cop_Gets__585K_After_Colleagues_Snooped_on_Her_DMV_Data___WIRED.pdf?1618168204                                                           51
https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pages/270/attachments/original/1618167029/KREKELBERG_v._Anoka_County__439_F._Supp._3d_1143_-_Dist._Court__Minnesota_2020_-_Google_Scholar.pdf?1618167029                           51
https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pages/270/attachments/original/1618167547/2016_USCOURTS-mnd-0_13-cv-03562-1.pdf?1618167547                                                                                         51
https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pages/270/attachments/original/1618166256/Krekelberg_Privacy_Lawsuit.pdf?1618166256                                                                                                50
https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pages/270/attachments/original/1614285311/2020_T

In [6203]:
unique_urls = sorted(list(unique_urls.index)) # get only the URLs

# All URLs seem valid, since they start with https
for url in unique_urls:
    print(url)

https://assets.nationbuilder.com/cuapb/pages/1462/attachments/original/1647659311/RCA-2022-00193_-_Legal_Settlement_Workers'_Compensation_claim_of_Jeffrey_Pennaz.pdf?1647659311
https://assets.nationbuilder.com/cuapb/pages/1462/attachments/original/1647659312/RCA-2022-00194_-_Legal_Settlement_Workers'_Compensation_claim_of_Kurt_Radke.pdf?1647659312
https://assets.nationbuilder.com/cuapb/pages/1462/attachments/original/1647659312/RCA-2022-00195_-_Legal_Settlement_Workers'_Compensation_claim_of_Craig_Taylor.pdf?1647659312
https://assets.nationbuilder.com/cuapb/pages/1462/attachments/original/1647659313/RCA-2022-00196_-_Legal_Settlement_Workers'_Compensation_claim_of_Cory_Taylor.pdf?1647659313
https://assets.nationbuilder.com/cuapb/pages/1462/attachments/original/1647659313/RCA-2022-00197_-_Legal_Settlement_Workers'_Compensation_claim_of_Joseph_Will.pdf?1647659313
https://assets.nationbuilder.com/cuapb/pages/1462/attachments/original/1647659314/RCA-2022-00198_-_Legal_Settlement_Workers'_Co

In [6204]:
starts_with_https = df["url"].str.startswith("https://") # Boolean series for whether each URL starts with "https"
sum(starts_with_https) == len(df["url"]) # Returns True if all the URLs in the dataframe start with https

True

### Identify duplicates

In [6205]:
is_duplicate = df.duplicated() # bool Series that is True if that row is a duplicate
sum(is_duplicate) # number of duplicated rows

17

In [6206]:
df[is_duplicate]

name  \
53              Andrew Allen, badge #37   
194         John E. Bennett, badge #406   
261           Peter Brazeau, badge #750   
302       Richard D. Bruns, badge #None   
368          Joseph Carroll, badge #200   
524      John C. Delmonico, badge #1491   
633            Mark Durand, badge #1627   
751        Michael Griffin, badge #2475   
961      Jeffrey R. Jindra, badge #3289   
1123        Joseph Klimmek, badge #3735   
1190   Joseph R. Labathe, badge #375760   
1314     Matthew Lindquist, badge #4169   
1700       George J. Peltz, badge #5546   
1933          Dylan Schultz, badge #208   
2012           Brian Sloat, badge #None   
2037         McCabe Stariha, badge #212   
2141  Jeffrey C. Thissen, badge #679500   

                                    department  \
53    Department:Minneapolis Police Department   
194   Department:Minneapolis Police Department   
261   Department:Minneapolis Police Department   
302        Department:Adrian Police Department   
368     Department:Richfield Police Department   
524   Department:Minneapolis Police Department   
633   Department:Minneapolis Police Department   
751   Department:Minneapolis Police Department   
961   Department:Minneapolis Police Department   
1123  Department:Minneapolis Police Department   
1190     Department:St. Paul Police Department   
1314  Department:Minneapolis Police Department   
1700  Department:Minneapolis Police Department   
1933    Department:Richfield Police Department   
2012  Department:Robbinsdale Police Department   
2037    Department:Richfield Police Department   
2141     Department:St. Paul Police Department   

                                                    url  \
53    https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
194   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
261   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
302   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
368   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
524   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
633   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
751   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
961   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1123  https://assets.nationbuilder.com/cuapb/pages/1...   
1190  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1314  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1700  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1933  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
2012  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
2037  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
2141  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   

                                                   text  
53    CASE 0:13-cv-00435-DSD-JJK   Document 1   File...  
194   3/28/2021 Errant bite ends Minneapolis police ...  
261   3/18/2021 Minneapolis officer ordered back on t...  
302   4/13/2021 Minnesota police officer arrested fo...  
368   3/6/2021 What we know about Brian Quinones, th...  
524   Minneapolis Police Officers:  350 S. 5th St.  ...  
633   4/11/2021 Cop Gets $585K After Colleagues Snoo...  
751   2/25/2021 Minneapolis cop under investigation ...  
961   3/2/2021 MPR: Feds won't prosecute Minneapolis...  
1123  From the desk of:  Medaria Arradondo  Chief of...  
1190  3/6/2021 What we know about Brian Quinones, th...  
1314  Too_LongPrint Date/Time:  Login ID:  Case Numb...  
1700  4/10/2021 Minneapolis pit bull shooting result...  
1933  3/6/2021 What we know about Brian Quinones, th...  
2012  4/8/2021 UPDATED: $185,000 settlement in Robbi...  
2037  3/6/2021 What we know about Brian Quinones, th...  
2141  MPR NEWS ›  Annotations   Police ID St. Louis ...

In [6207]:
dup_indexes = df[is_duplicate].index # indices of duplicate rows

for index in dup_indexes:
    print(f"\n{'-'*100}\n{index}: URL: {df["url"][index]}\n\n{df["text"][index]}")


----------------------------------------------------------------------------------------------------
53: URL: https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pages/270/attachments/original/1614808522/2015DanielFancherComplaint.pdf?1614808522

CASE 0:13-cv-00435-DSD-JJK   Document 1   Filed 02/22/13   Page 1 of 11 CASE 0:13-cv-00435-DSD-JJK   Document 1   Filed 02/22/13   Page 2 of 11 CASE 0:13-cv-00435-DSD-JJK   Document 1   Filed 02/22/13   Page 3 of 11 CASE 0:13-cv-00435-DSD-JJK   Document 1   Filed 02/22/13   Page 4 of 11 CASE 0:13-cv-00435-DSD-JJK   Document 1   Filed 02/22/13   Page 5 of 11 CASE 0:13-cv-00435-DSD-JJK   Document 1   Filed 02/22/13   Page 6 of 11 CASE 0:13-cv-00435-DSD-JJK   Document 1   Filed 02/22/13   Page 7 of 11 CASE 0:13-cv-00435-DSD-JJK   Document 1   Filed 02/22/13   Page 8 of 11 CASE 0:13-cv-00435-DSD-JJK   Document 1   Filed 02/22/13   Page 9 of 11 CASE 0:13-cv-00435-DSD-JJK   Document 1   Filed 02/22/13   Page 10 of 11 CASE 0:13-cv-00435-DSD-JJK   Document 1

We will drop the duplicate rows.

In [6208]:
df = df.drop_duplicates(ignore_index=True) # remove duplicate rows and reset the index to a (new) sequential integer index
len(df)

2390

### Check invalid values for text

In [6209]:
df.sort_values(by = "text")

name  \
1809    Peter J. Ritschel, badge #6037   
1867    Peter J. Ritschel, badge #6037   
249    Christopher Bennett, badge #413   
149           Tyrone Barze, badge #330   
1444      Phillip Meemken, badge #None   
...                                ...   
1829    Peter J. Ritschel, badge #6037   
781       Michael Griffin, badge #2475   
639      David C. Elliott, badge #1755   
362   Jason Allen Carlson, badge #None   
359   Jason Allen Carlson, badge #None   

                                          department  \
1809        Department:Minneapolis Police Department   
1867        Department:Minneapolis Police Department   
249         Department:Minneapolis Police Department   
149         Department:Minneapolis Police Department   
1444  Department:Stearns County Sheriff's Department   
...                                              ...   
1829        Department:Minneapolis Police Department   
781         Department:Minneapolis Police Department   
639         Department:Minneapolis Police Department   
362                 Department:Ely Police Department   
359                 Department:Ely Police Department   

                                                    url  \
1809  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1867  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
249   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
149   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1444  https://assets.nationbuilder.com/cuapb/pages/1...   
...                                                 ...   
1829  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
781   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
639   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
362   https://assets.nationbuilder.com/cuapb/pages/1...   
359   https://assets.nationbuilder.com/cuapb/pages/1...   

                                                   text  
1809      


      ...  
1867  "bad faith" has been characterized as conduct ...  
249   ()  At this point the discrepancies between th...  
149   () Mpls. pays $82K in case of Maple Grove woma...  
1444  (/)  (/submit/)  (https://twitter.com/news8new...  
...                                                 ...  
1829  • Ritschel took Dahl's bike and brought it aro...  
781   • Sgt. Biederman was called to the scene and c...  
639   •Officer Elliott accessed 738 photos through E...  
362    Support the Timberjay by making a donation. ...  
359    Support the Timberjay by making a donation. ...  

[2390 rows x 4 columns]

Below we check all rows that seem to have invalid text.

In [6210]:
print(f"Text: {df.copy().loc[1809, "text"]}")
print(f"URL: {df.copy().loc[1809, "url"]}")

df = df.drop(index=1809) # Drop row since it has a bunch of gibberish characters

Text:                   !"#$%"#$ &"'$() "'$*& !+  , !$(*'   -./0./12
URL: https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pages/270/attachments/original/1626475811/Ritschel_hearing.pdf?1626475811


In [6211]:
df.loc[1444, "text"] # Keep this row since it's a valid article, just with a bunch of URLs at the start

'(/)  (/submit/)  (https://twitter.com/news8news)  (https://www.facebook.com/news8000?ref=hl)  (https://www.instagram.com/wkbtnews8/?hl=en)  (https://www.youtube.com/user/wkbttv) Plea deal for Minnesota deputy accused of sexual abuse Posted: May 7, 2012 9:07 PM Updated: December 21, 2019 10:17 AM by Site sta\x01 (https://www.news8000.com/author/sitesta\x01/) ANOKA, Minn. (AP) — A Stearns County sheri\x01’s deputy has reached a plea deal on charges that he sexually abused three teenagers. The St. Cloud Times reports Sgt. Phil Meemken reached the deal Monday as jury selection was scheduled to begin for his trial in Anoka County. Paul Young, head of the violent crime division of the Anoka County Attorney’s O\x02ce, tells the newspaper Meemken will plead guilty to one count of fourth-degree criminal sexual conduct and two counts of providing alcohol to a minor. Meemken was facing 25 counts, Including 22 counts of criminal sexual conduct. He was accused of providing alcohol to three teenage

In [6212]:
df.loc[362, "text"] # This one is valid. But the article seems to be incomplete because it has 2 pages out of 5

"\uf071 Support the Timberjay by making a donation.  (/supportus/ ) Welcome! \xa0  Log in (/) Serving Northern St. Louis County, Minnesota \uf071You have 2 free items remaining before a subscription is required. \xa0 Subscribe now! (/subscribe/?town_id=0) \xa0 | \xa0 Log in (/login.html) Carlson resigns as Ely police o\x01cer City’s employee committee recommends denial, urges termination instead Posted Wednesday, February 1, 2017 4:15 pm Keith Vandervort ELY - Ely police o\x01cer Jason Carlson, who avoided jail time after striking a plea agreement in court and admitting to an illegal sexual relationship with an underage girl, submitted his resignation to the city o\x01cials last week. The city’s employee relations committee recommended that the City Council deny the request. The Council will discuss and could take action on the issue at their next meeting on Feb. 7. Carlson, an 11-year veteran of the city’s police force, was on paid administrative leave with the city from April 2015 un

Below we check the length of all strings in the text column, to see if any strings are empty or very short.

In [6213]:
# Add a column for the length of each string in the text column, just for data cleaning purposes. We will remove it after data cleaning.
df["text_length"] = df["text"].str.len()

for length in sorted(list(df["text_length"])): # print the string length of each entry, in ascending order
    print(length)

13
13
13
13
13
13
13
13
13
13
13
13
13
13
13
13
23
71
85
92
100
100
143
143
143
143
143
143
143
143
143
143
143
143
143
143
143
143
143
167
175
176
189
189
189
189
190
190
190
190
190
191
191
191
191
191
191
191
191
192
192
192
192
192
192
192
192
192
192
192
192
192
192
192
192
192
192
192
192
192
192
192
192
192
193
193
195
214
215
216
216
216
217
218
218
218
218
218
218
218
219
222
224
225
225
231
272
283
283
283
283
283
344
354
355
355
357
378
378
379
379
379
379
379
379
379
380
380
381
381
381
381
381
383
385
385
385
385
385
385
385
385
385
385
385
385
385
385
385
385
385
389
389
407
407
407
407
407
407
407
407
407
418
420
425
461
496
496
496
496
567
567
567
575
575
575
575
575
575
575
575
575
575
575
575
575
575
575
575
575
578
580
580
596
596
658
680
697
698
701
711
711
711
711
711
750
777
777
783
783
793
793
794
818
818
818
818
818
818
818
830
880
881
881
890
893
895
898
899
914
923
939
939
939
939
939
939
939
959
959
959
959
959
959
959
959
959
959
975
975
982
999
999
999
999


In [6214]:
df.head()

name  \
0      Andrew Allen, badge #37   
1  Guled Abdullahi, badge #706   
2  Dean V. Albers, badge #None   
3      Scott Aikins, badge #22   
4    Matthew Aish, badge #None   

                                        department  \
0         Department:Minneapolis Police Department   
1  Department:Hennepin County Sheriff's Department   
2   Department:Goodhue County Sheriff's Department   
3         Department:Minneapolis Police Department   
4    Department:Columbia Heights Police Department   

                                                 url  \
0  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1  https://assets.nationbuilder.com/cuapb/pages/1...   
2  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
3  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
4  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   

                                                text  text_length  
0  Home Legislative File 2021-01132 RCA Legal s...         1426  
1  Hennepin County  300 South Sixth Street, Minne...         1476  
2  2/22/2021 Jenson v. Craft, Civil No. 01-1488(D...        34621  
3  2/22/2021 United States v. Diriye, Case No. 14...        22890  
4  Too_Long1    Arbitration   LELS (Mathew Aish)/...        44726

Some strings have only 13 characters, which is very few for an article. Let's check those:

In [6215]:
rows_13 = df.loc[df["text_length"] == 13] # all data rows where text length is 13
rows_13

name  \
36        Brian Anderson, badge #91   
186      Justin Benner, badge #None   
333        Charles Cape, badge #971   
369     Timothy Carson, badge #1006   
684         Joe Fuller, badge #2160   
707     Matthew George, badge #None   
1412     Cullan McHarg, badge #None   
1442       Nick Melser, badge #None   
1524    Justin G Mullen, badge #122   
1832      David Robins, badge #6046   
2107  Lafayette Temple, badge #None   
2109  Lafayette Temple, badge #None   
2154     Clifton Toles, badge #7243   
2234  Katherine Walker, badge #None   
2240     Daniel Walton, badge #None   
2318       Jason Wolff, badge #7859   

                                      department  \
36      Department:Minneapolis Police Department   
186    Department:Ramsey County Sheriff's Office   
333     Department:Minneapolis Police Department   
369     Department:Minneapolis Police Department   
684     Department:Minneapolis Police Department   
707     Department:Bloomington Police Department   
1412    Department:Bloomington Police Department   
1442    Department:Bloomington Police Department   
1524     Department:Burnsville Police Department   
1832    Department:Minneapolis Police Department   
2107  Department:Metro Transit Police Department   
2109  Department:Metro Transit Police Department   
2154    Department:Minneapolis Police Department   
2234    Department:Bloomington Police Department   
2240    Department:Bloomington Police Department   
2318    Department:Minneapolis Police Department   

                                                    url           text  \
36    https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Access denied   
186   https://assets.nationbuilder.com/cuapb/pages/1...  Access denied   
333   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Access denied   
369   https://assets.nationbuilder.com/cuapb/pages/1...  Access denied   
684   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Access denied   
707   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Access denied   
1412  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Access denied   
1442  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Access denied   
1524  https://assets.nationbuilder.com/cuapb/pages/1...  Access denied   
1832  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Access denied   
2107  https://assets.nationbuilder.com/cuapb/pages/1...  Access denied   
2109  https://assets.nationbuilder.com/cuapb/pages/1...  Access denied   
2154  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Access denied   
2234  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Access denied   
2240  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Access denied   
2318  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...  Access denied   

      text_length  
36             13  
186            13  
333            13  
369            13  
684            13  
707            13  
1412           13  
1442           13  
1524           13  
1832           13  
2107           13  
2109           13  
2154           13  
2234           13  
2240           13  
2318           13

Since all of these rows just say "Access denied", we will drop them.

In [6216]:
df.drop(index = rows_13.index, inplace = True) # drop invalid rows

In [6217]:
sorted(list(df["text_length"]))

[23,
 71,
 85,
 92,
 100,
 100,
 143,
 143,
 143,
 143,
 143,
 143,
 143,
 143,
 143,
 143,
 143,
 143,
 143,
 143,
 143,
 143,
 143,
 167,
 175,
 176,
 189,
 189,
 189,
 189,
 190,
 190,
 190,
 190,
 190,
 191,
 191,
 191,
 191,
 191,
 191,
 191,
 191,
 192,
 192,
 192,
 192,
 192,
 192,
 192,
 192,
 192,
 192,
 192,
 192,
 192,
 192,
 192,
 192,
 192,
 192,
 192,
 192,
 192,
 192,
 192,
 192,
 192,
 193,
 193,
 195,
 214,
 215,
 216,
 216,
 216,
 217,
 218,
 218,
 218,
 218,
 218,
 218,
 218,
 219,
 222,
 224,
 225,
 225,
 231,
 272,
 283,
 283,
 283,
 283,
 283,
 344,
 354,
 355,
 355,
 357,
 378,
 378,
 379,
 379,
 379,
 379,
 379,
 379,
 379,
 380,
 380,
 381,
 381,
 381,
 381,
 381,
 383,
 385,
 385,
 385,
 385,
 385,
 385,
 385,
 385,
 385,
 385,
 385,
 385,
 385,
 385,
 385,
 385,
 385,
 389,
 389,
 407,
 407,
 407,
 407,
 407,
 407,
 407,
 407,
 407,
 418,
 420,
 425,
 461,
 496,
 496,
 496,
 496,
 567,
 567,
 567,
 575,
 575,
 575,
 575,
 575,
 575,
 575,
 575,
 575,
 575,
 5

Now check all rows where the text has 100 or less characters, since that is still very few for an article.

In [6218]:
rows_100 = df.loc[df["text_length"] <= 100]
rows_100

name  \
106            Julio Baez, badge #None   
744       Michael Griffin, badge #2475   
1350       Michael Manley, badge #4385   
1558   Patrick V. McMahon, badge #4636   
2276  Christopher Steward, badge #6823   
2372         Michael Zueg, badge #None   

                                     department  \
106         Department:Kasson Police Department   
744    Department:Minneapolis Police Department   
1350   Department:Minneapolis Police Department   
1558   Department:Minneapolis Police Department   
2276   Department:Minneapolis Police Department   
2372  Department:Walnut Grove Police Department   

                                                    url  \
106   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
744   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1350  https://assets.nationbuilder.com/cuapb/pages/1...   
1558  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
2276  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
2372  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   

                                                   text  text_length  
106   https://minnesota.cbslocal.com/2018/06/21/kass...           92  
744   CASE 0:12-cv-02551-DWF-JSM   Document 15   Fil...           71  
1350                            1336323 1336324 1336325           23  
1558  MR. DENNIS RODOLFO HERNANDEZ-ANTUNEZ  1147 N 7...          100  
2276  MR. DENNIS RODOLFO HERNANDEZ-ANTUNEZ  1147 N 7...          100  
2372  https://minnesota.cbslocal.com/2018/07/25/poli...           85

In [6219]:
for index in rows_100.index:
    print(f"{index}: {rows_100.loc[index, "text"]}")

106: https://minnesota.cbslocal.com/2018/06/21/kasson-pd-officer-accused-criminal-sexual-conduct/
744: CASE 0:12-cv-02551-DWF-JSM   Document 15   Filed 01/26/15   Page 1 of 1
1350: 1336323 1336324 1336325
1558: MR. DENNIS RODOLFO HERNANDEZ-ANTUNEZ  1147 N 7TH ST  APT E-3041  MINNEAPOLIS, MN  55411  CUAPB002843
2276: MR. DENNIS RODOLFO HERNANDEZ-ANTUNEZ  1147 N 7TH ST  APT E-3041  MINNEAPOLIS, MN  55411  CUAPB002843
2372: https://minnesota.cbslocal.com/2018/07/25/police-chief-pleads-guilty-to-prostitution/


All these rows do not contain an article, so we will drop them.

In [6220]:
df.drop(index = rows_100.index, inplace=True)
len(df)

2367

Now check text with 200 characters or less.

In [6221]:
# Get rows with length <= 200, and sort the dataframe in ascending order by the length of the text.
rows_200 = df.loc[df["text_length"] <= 200].sort_values(by = "text_length")
rows_200

name  \
1157         Robert J. Kroll, badge #3874   
2169        Geoffrey Toscano, badge #7257   
2003     Donald Smulski, Jr., badge #6696   
1999        Katherine Smulski, badge #468   
1987          Roger D. Smith, badge #6689   
...                                   ...   
1646  Brett R. Palkowitsch, badge #520305   
85         Heather J. Aschoff, badge #233   
538          Joshua Demmerly, badge #None   
382             Tou M. Cha, badge #112410   
2241           Kent Warnberg, badge #7570   

                                    department  \
1157  Department:Minneapolis Police Department   
2169  Department:Minneapolis Police Department   
2003  Department:Minneapolis Police Department   
1999  Department:Minneapolis Police Department   
1987  Department:Minneapolis Police Department   
...                                        ...   
1646     Department:St. Paul Police Department   
85    Department:Minneapolis Police Department   
538       Department:Warroad Police Department   
382      Department:St. Paul Police Department   
2241  Department:Minneapolis Police Department   

                                                    url  \
1157  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
2169  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
2003  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1999  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1987  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
...                                                 ...   
1646  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
85    https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
538   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
382   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
2241  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   

                                                   text  text_length  
1157  CASE 0:03-cv-01182-DSD-SRN   Document 16   Fil...          143  
2169  CASE 0:06-cv-00579-DWF-AJB   Document 51   Fil...          143  
2003  CASE 0:06-cv-00579-DWF-AJB   Document 51   Fil...          143  
1999  CASE 0:06-cv-00579-DWF-AJB   Document 51   Fil...          143  
1987  CASE 0:06-cv-00579-DWF-AJB   Document 51   Fil...          143  
...                                                 ...          ...  
1646  / Star Tribune (Minneapolis, Minnesota) ·  Thu...          192  
85    Star Tribune (Minneapolis, Minnesota) ·  Wed, ...          192  
538   Star Tribune (Minneapolis, Minnesota) ·  Sun, ...          193  
382   Star Tribune (Minneapolis, Minnesota) ·  Fri, ...          193  
2241  The Winona Daily News (Winona, Minnesota) ·  F...          195  

[65 rows x 5 columns]

In [6222]:
for index in rows_200.index:
    print(f"{'-'*50}\n{index}: {rows_200.loc[index, "text"]}\n")

--------------------------------------------------
1157: CASE 0:03-cv-01182-DSD-SRN   Document 16   Filed 04/13/04   Page 1 of 2 CASE 0:03-cv-01182-DSD-SRN   Document 16   Filed 04/13/04   Page 2 of 2

--------------------------------------------------
2169: CASE 0:06-cv-00579-DWF-AJB   Document 51   Filed 10/19/07   Page 1 of 2 CASE 0:06-cv-00579-DWF-AJB   Document 51   Filed 10/19/07   Page 2 of 2

--------------------------------------------------
2003: CASE 0:06-cv-00579-DWF-AJB   Document 51   Filed 10/19/07   Page 1 of 2 CASE 0:06-cv-00579-DWF-AJB   Document 51   Filed 10/19/07   Page 2 of 2

--------------------------------------------------
1999: CASE 0:06-cv-00579-DWF-AJB   Document 51   Filed 10/19/07   Page 1 of 2 CASE 0:06-cv-00579-DWF-AJB   Document 51   Filed 10/19/07   Page 2 of 2

--------------------------------------------------
1987: CASE 0:06-cv-00579-DWF-AJB   Document 51   Filed 10/19/07   Page 1 of 2 CASE 0:06-cv-00579-DWF-AJB   Document 51   Filed 10/19/07   Pag

Double-check rows I'm unsure about:

In [6223]:
print(df.loc[330, "text"])

Minneapolis police officer jailed, accused of sex crime - StarTribune.com http://www.startribune.com/minneapolis-police-officer-jailed-accused-of... 1 of 1 9/10/2017, 12:24 PM


In [6224]:
df.loc[144, "text"]

'Minneapolis officer accused of strong-arm tactics with canvasser - StarT... http://www.startribune.com/minneapolis-officer-accused-of-strong-arm-ta... 1 of 1 8/25/2018, 2:37 AM'

We will drop all of these rows. None of them contain articles, and most of them just contain the title and URL of an article.

In [6225]:
df.drop(index=rows_200.index, inplace=True)
len(df)

2302

In [6226]:
unique_text_lengths = df["text_length"].value_counts() # unique values of text length
unique_text_lengths

text_length
72671    51
38274    51
7970     50
33017    50
53556    18
         ..
7914      1
3274      1
4346      1
2375      1
42150     1
Name: count, Length: 1248, dtype: int64

In [6227]:
for length in list(unique_text_lengths.values):
    print(length)

51
51
50
50
18
17
17
17
15
15
15
14
14
11
10
10
10
10
9
9
8
8
8
7
7
7
7
7
7
7
7
7
7
7
7
6
6
6
6
6
6
6
6
6
6
6
6
6
5
5
5
5
5
5
5
5
5
5
5
5
5
5
5
5
5
5
5
5
5
5
5
5
5
4
4
4
4
4
4
4
4
4
4
4
4
4
4
4
4
4
4
4
4
4
4
4
4
4
4
4
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
3
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1


In [6228]:
for l in sorted(list(df["text_length"])):
    print(l)

214
215
216
216
216
217
218
218
218
218
218
218
218
219
222
224
225
225
231
272
283
283
283
283
283
344
354
355
355
357
378
378
379
379
379
379
379
379
379
380
380
381
381
381
381
381
383
385
385
385
385
385
385
385
385
385
385
385
385
385
385
385
385
385
389
389
407
407
407
407
407
407
407
407
407
418
420
425
461
496
496
496
496
567
567
567
575
575
575
575
575
575
575
575
575
575
575
575
575
575
575
575
575
578
580
580
596
596
658
680
697
698
701
711
711
711
711
711
750
777
777
783
783
793
793
794
818
818
818
818
818
818
818
830
880
881
881
890
893
895
898
899
914
923
939
939
939
939
939
939
939
959
959
959
959
959
959
959
959
959
959
975
975
982
999
999
999
999
1010
1027
1027
1027
1027
1027
1027
1027
1027
1028
1038
1038
1038
1038
1038
1038
1077
1079
1079
1079
1079
1079
1079
1079
1089
1093
1099
1101
1104
1104
1115
1115
1115
1117
1117
1120
1120
1120
1120
1127
1136
1151
1151
1161
1161
1161
1161
1173
1180
1180
1180
1180
1180
1180
1198
1198
1198
1207
1218
1238
1247
1247
1247
1247
1248
125

In [6229]:
len(df["text"].unique()), len(df) # check for duplicate texts

(1394, 2302)

There are 2,319 texts but only 1,394 unique texts. Let's check the duplicates.

In [6230]:
is_duplicate_text = df["text"].duplicated() # bool series that is True if the text in a row has a duplicate. Mark all duplicates, not just the first occurrence.
df["is_text_duplicated"] = is_duplicate_text # add a column for whether a row has a duplicate
sum(is_duplicate_text), len(df)

(908, 2302)

In [6231]:
df[is_duplicate_text].sort_values(by = "text").head(20)

name  \
293        Jeanine Brudenell, badge #845   
2277         Roderick Weber, badge #7612   
1466      Timothy J. Merkel, badge #4744   
734            Robert Greer, badge #2441   
422       Erika Christensen, badge #1097   
895     Timothy J. Hoeppner, badge #3051   
1746      Daniel Pommerenke, badge #5774   
1555       Edward T. Nelson, badge #4970   
210       John A. Billington, badge #556   
1078          Joel Kimmerle, badge #3703   
1020           Vicki Karnik, badge #3637   
1462          Justin Merten, badge #4752   
1385       Michael McCarthy, badge #4520   
955         James R. Jensen, badge #3282   
626      Andrew P. Enriquez, badge #1802   
849            Lupe Herrera, badge #2973   
722   Valerie S. Goligowski, badge #2312   
564            Hien Q. Dinh, badge #1551   
1355         Steven Manhood, badge #4323   
48          Mohamed Abdullahi, badge #28   

                                    department  \
293   Department:Minneapolis Police Department   
2277  Department:Minneapolis Police Department   
1466  Department:Minneapolis Police Department   
734   Department:Minneapolis Police Department   
422   Department:Minneapolis Police Department   
895   Department:Minneapolis Police Department   
1746  Department:Minneapolis Police Department   
1555  Department:Minneapolis Police Department   
210   Department:Minneapolis Police Department   
1078  Department:Minneapolis Police Department   
1020  Department:Minneapolis Police Department   
1462  Department:Minneapolis Police Department   
1385  Department:Minneapolis Police Department   
955   Department:Minneapolis Police Department   
626   Department:Minneapolis Police Department   
849   Department:Minneapolis Police Department   
722   Department:Minneapolis Police Department   
564   Department:Minneapolis Police Department   
1355  Department:Minneapolis Police Department   
48    Department:Minneapolis Police Department   

                                                    url  \
293   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
2277  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1466  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
734   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
422   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
895   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1746  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1555  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
210   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1078  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1020  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1462  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1385  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
955   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
626   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
849   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
722   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
564   https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1355  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
48    https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   

                                                   text  text_length  \
293   (https://lp.¡ndlaw.com/) FINDLAW (HTTPS://LP.F...        78719   
2277  (https://lp.¡ndlaw.com/) FINDLAW (HTTPS://LP.F...        78719   
1466  (https://lp.¡ndlaw.com/) FINDLAW (HTTPS://LP.F...        78719   
734   (https://lp.¡ndlaw.com/) FINDLAW (HTTPS://LP.F...        78719   
422   (https://lp.¡ndlaw.com/) FINDLAW (HTTPS://LP.F...        78719   
895   (https://lp.¡ndlaw.com/) FINDLAW (HTTPS://LP.F...        78719   
1746  (https://lp.¡ndlaw.com/) FINDLAW (HTTPS://LP.F...        78719   
1555  (https://lp.¡ndlaw.com/) FINDLAW (HTTPS://LP.F...        78719   
210   (https://lp.¡ndlaw.com/) FINDLAW (HTTPS://LP.F...        78719   
1078  (https://lp.¡ndlaw.com/) FINDLAW (HTTPS://LP.F...        19746   
1020  (https://lp.¡ndlaw.com

In [6232]:
sum(df[["text", "url"]].duplicated()) # how many rows have the same text and URL as another row

900

So there are 8 rows that have the same text as another row, but a different URL from the other row.

In [6233]:
sum(df[["text", "url", "name"]].duplicated()) # how many rows have the same text, URL, AND officer name as another row. There are no such rows.

0

Problem: If an article is about multiple police officers, then each officer gets their own row, and all those rows contain the same article text.

### Check invalid `name` values

In [6234]:
for name in sorted(list(df["name"].unique())):
    print(name)

Aaron Apitz, badge #None
Aaron C. Morrison, badge #4859
Aaron Collins, badge #1259
Aaron Heuer, badge #None
Aaron L. Biard, badge #542
Aaron Pearson, badge #5504
Aaron Prescott, badge #5806
Aaron Womble, badge #7851
Abdiaziz Omar, badge #None
Abdirashid Mohamed, badge #None
Abdiwahab Ali, badge #30
Abdulkhayr  Hirse, badge #None
Abraham T. Cyr, badge #139000
Abubakar Muridi, badge #4896
Adam Castilleja, badge #None
Adam Hakanson, badge #3069
Adam Lepinski, badge #4093
Adam Lewis, badge #4102
Adam Miller, badge #659
Adam Moen, badge #4832
Adan Casas-Uriostegui, badge #None
Aimee Walker (Linson), badge #4172
Aine M. Bebeau, badge #35000
Alan K. Nielsen, badge #None
Alan Knowlton, badge #None
Alan Liotta, badge #4196
Alan Salvosa, badge #None
Alan W. Williams, badge #7763
Alberto Mendez, badge #509
Alex Nonnemacher, badge #5157
Alexander Brown, badge #820
Alexander Graham, badge #510
Alexander Walls, badge #7498
Alexandra Dubay, badge #1617
Alexandra Urgiles, badge #685
Allan Kertscher, b

All names look valid.

In [6235]:
name_lengths = sorted(list(df["name"].str.len())) # check for empty strings in the name column. There are none

for l in name_lengths:
    print(l)

20
20
20
21
21
21
21
21
21
21
21
21
21
21
21
21
21
21
21
21
21
21
21
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
22
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
23
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
24
2

### Check invalid departments

In [6236]:
sorted(list(df["department"].unique())) # all unique department names

['Department:',
 'Department:Adams Police Department',
 'Department:Adrian Police Department',
 'Department:Alexandria Police Department',
 "Department:Anoka County Sheriff's Office",
 'Department:Babbitt Police Department',
 'Department:Barnesville Police Department',
 'Department:Bayport Police Department',
 'Department:Belle Plaine Police Department',
 'Department:Bloomington Police Department',
 'Department:Braham Police Department',
 'Department:Brooklyn Center Police Department',
 'Department:Brooklyn Park Police Department',
 'Department:Brownsdale Police Department',
 'Department:Bureau of Criminal Apprehension',
 'Department:Burnsville Police Department',
 'Department:Canby Police Department',
 "Department:Carlton County Sheriff's Department",
 "Department:Carver County Sheriff's Department",
 'Department:Champlin Police Department',
 'Department:Chaska Police Department',
 'Department:Chatfield Police Department',
 "Department:Chisago County Sheriff's Department",
 'Departmen

Some rows have a `department` value of just `Department:`, so we will look at those.

In [6237]:
no_dept_rows = df.loc[df["department"] == "Department:"] # rows with no department specified
no_dept_rows

name   department  \
371   Adan Casas-Uriostegui, badge #None  Department:   
531             Brian Denny, badge #None  Department:   
1116          Alan Knowlton, badge #None  Department:   
1125          Alan Knowlton, badge #None  Department:   
1203        David W. Larson, badge #None  Department:   
1470      Richard G. Miller, badge #None  Department:   
1680        Shawn L. Peters, badge #None  Department:   
1857            Curtis Rude, badge #None  Department:   
2364           Michael Zeug, badge #None  Department:   

                                                    url  \
371   https://assets.nationbuilder.com/cuapb/pages/1...   
531   https://assets.nationbuilder.com/cuapb/pages/1...   
1116  https://assets.nationbuilder.com/cuapb/pages/1...   
1125  https://assets.nationbuilder.com/cuapb/pages/1...   
1203  https://assets.nationbuilder.com/cuapb/pages/1...   
1470  https://assets.nationbuilder.com/cuapb/pages/1...   
1680  https://assets.nationbuilder.com/cuapb/pages/1...   
1857  https://assets.nationbuilder.com/cuapb/pages/1...   
2364  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   

                                                   text  text_length  \
371   C C STATE OF MINNESOTA BOARD OF PEACE OFFICER ...         2416   
531   C C STATE OF MINNESOTA BOARD OF PEACE OFFICER ...         2522   
1116  C C) STATE OF MINNESOTA BOARD OF PEACE OFFICER...         3836   
1125  Skip to Main Content Logout My Account Search ...         4045   
1203  C C STATE OF MINNEOSTA BOARD OF PEACE OFFICER ...         2419   
1470  Skip to Main Content Logout My Account Search ...         4265   
1680  Skip to Main Content Logout My Account Search ...         2278   
1857  C C S ‘ STATE OF MINNESOTA BOARD OF PEACE OFFi...         2373   
2364  STAR TRIBUNE ›  Annotations   Off-duty McLeod ...         1403   

      is_text_duplicated  
371                False  
531                False  
1116               False  
1125               False  
1203               False  
1470               False  
1680               False  
1857               False  
2364                True

In [6238]:
for index in no_dept_rows.index:
    print(f"\n{'-'*100}\n{index}: URL: {df.loc[index, "url"]}\n\n{df.loc[index, "text"]}")


----------------------------------------------------------------------------------------------------
371: URL: https://assets.nationbuilder.com/cuapb/pages/1472/attachments/original/1669104724/Casas-Uriostegui_Adan-20-039-13919.pdf?1669104724

C C STATE OF MINNESOTA BOARD OF PEACE OFFICER STANDARDS AND TRAINING In the Matter of the Peace Officer License of Adan Casas-Uriostegui License Number: 13919 1. The Minnesota Board of Peace Officer Standards and Training (“Board”) is authorized pursuant to Minnesota Statutes sections 626.84 through 626.90 (1998) and Minnesota Rules chapter 6700 to license, regulate, and discipline persons who apply for, petition, or hold peace officer licenses in the State of Minnesota and is further authorized pursuant to Minnesota Statutes section 214.10 to review complaints against peace officers and to initiate appropriate disciplinary action. 2. Adan Casas-Uriostegui (“Respondent”) has been a peace officer in Minnesota since June 18, 1997. 3. Pursuant to M

All these rows look valid, so we will set their `department` value to `NaN` instead of dropping them.

In [6239]:
df["department"].dtype # check that the data types are the same before and after

dtype('O')

In [6240]:
df.loc[no_dept_rows.index, "department"] = None
df.loc[df["department"].isna()]

name department  \
371   Adan Casas-Uriostegui, badge #None       None   
531             Brian Denny, badge #None       None   
1116          Alan Knowlton, badge #None       None   
1125          Alan Knowlton, badge #None       None   
1203        David W. Larson, badge #None       None   
1470      Richard G. Miller, badge #None       None   
1680        Shawn L. Peters, badge #None       None   
1857            Curtis Rude, badge #None       None   
2364           Michael Zeug, badge #None       None   

                                                    url  \
371   https://assets.nationbuilder.com/cuapb/pages/1...   
531   https://assets.nationbuilder.com/cuapb/pages/1...   
1116  https://assets.nationbuilder.com/cuapb/pages/1...   
1125  https://assets.nationbuilder.com/cuapb/pages/1...   
1203  https://assets.nationbuilder.com/cuapb/pages/1...   
1470  https://assets.nationbuilder.com/cuapb/pages/1...   
1680  https://assets.nationbuilder.com/cuapb/pages/1...   
1857  https://assets.nationbuilder.com/cuapb/pages/1...   
2364  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   

                                                   text  text_length  \
371   C C STATE OF MINNESOTA BOARD OF PEACE OFFICER ...         2416   
531   C C STATE OF MINNESOTA BOARD OF PEACE OFFICER ...         2522   
1116  C C) STATE OF MINNESOTA BOARD OF PEACE OFFICER...         3836   
1125  Skip to Main Content Logout My Account Search ...         4045   
1203  C C STATE OF MINNEOSTA BOARD OF PEACE OFFICER ...         2419   
1470  Skip to Main Content Logout My Account Search ...         4265   
1680  Skip to Main Content Logout My Account Search ...         2278   
1857  C C S ‘ STATE OF MINNESOTA BOARD OF PEACE OFFi...         2373   
2364  STAR TRIBUNE ›  Annotations   Off-duty McLeod ...         1403   

      is_text_duplicated  
371                False  
531                False  
1116               False  
1125               False  
1203               False  
1470               False  
1680               False  
1857               False  
2364                True

In [6241]:
df["department"].dtype

dtype('O')

In [6242]:
# Removes the "Department:" prefix
df["department"] = df["department"].str.removeprefix("Department:")
df.head()

name                            department  \
0      Andrew Allen, badge #37         Minneapolis Police Department   
1  Guled Abdullahi, badge #706  Hennepin County Sheriff's Department   
2  Dean V. Albers, badge #None   Goodhue County Sheriff's Department   
3      Scott Aikins, badge #22         Minneapolis Police Department   
4    Matthew Aish, badge #None    Columbia Heights Police Department   

                                                 url  \
0  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
1  https://assets.nationbuilder.com/cuapb/pages/1...   
2  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
3  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   
4  https://d3n8a8pro7vhmx.cloudfront.net/cuapb/pa...   

                                                text  text_length  \
0  Home Legislative File 2021-01132 RCA Legal s...         1426   
1  Hennepin County  300 South Sixth Street, Minne...         1476   
2  2/22/2021 Jenson v. Craft, Civil No. 01-1488(D...        34621   
3  2/22/2021 United States v. Diriye, Case No. 14...        22890   
4  Too_Long1    Arbitration   LELS (Mathew Aish)/...        44726   

   is_text_duplicated  
0               False  
1               False  
2               False  
3               False  
4               False

In [6243]:
df.drop(columns = ["text_length", "is_text_duplicated"]).to_csv("cleaned_police_reports.csv", index=False)